enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+Also to support kubevirt live migration the GR LRP will have an extra address with the configured gateway for the layer2 subnet so the gateway IP is independent of the node where the VM is running on, after live migration OVN should send and GARP for VMs to cleanup its ARP tables since the gateway IP has different mac now.
+
+
+
+Following is a picture that illustrate all these bits with a topology
+
+
+![Layer 2 Egress Topology](images/multi-homing-l2-gw.svg)
+
+
+
+#### Services
+
+When ingress service traffic enters br-ex, there are flows installed that steer service traffic towards the OVN GR. With additional networks, these flows will be modified to steer traffic to the correct GR-&lt;network>’s patch port.
+
+When a host process or host networked pod on a Kubernetes node initiates a connection to a service, iptables rules will DNAT the nodeport or loadbalancer IP into the cluster IP, and then send the traffic via br-ex where it is masqueraded and sent into the OVN GR. These flows can all be modified to detect the service IP and then send to the correct GR-&lt;network> patch port. However, in the case where there is hairpin service traffic to the host (Host->Service->Endpoint is also the host), the endpoint reply traffic will need to be distinguishable on a per network basis. In order to achieve this, each OVN GR’s unique masquerade IP will be leveraged.
+
+For service access towards KAPI/DNS or potentially other services on the default network, there are two potential technical solutions. Assume eth0 is the pod interface connected to the default network, and eth1 is connected to the unique primary network:
+
+1. Add routes for KAPI/DNS specifically into the pod to go out eth0, while all other service access will go to eth1. This will then just work normally with the load balancers on the switches for the respective networks.
+
+2. Do not send any service traffic out of eth0, instead all service traffic goes to eth1. ovn_cluster_router_&lt;unique network> is configured to send service CIDR traffic to ovn-k8s-mp0-&lt;unique network>. ipTables rules in the host only permit access to KAPI/DNS and drop all other service traffic coming from ovn-k8s-mp0-&lt;unique network>. The traffic then gets routed to br-ex and default GR where it hits the service.
+
+While the second option is more complex, it allows for not configuring routes to service addresses in the pod that could hypothetically change.
```

Comment:
on a separate discussion you've told me of yet another issue w/ option 1:  if we want to allow access to all services on the default network then we certainly would need to dynamically change routing as those services change. 

Wouldn't we be able to just write a route for the entire service network ?... For instance, this is what gets configured *today* if I configure a pod's default route to be on a **secondary** network. OVN-K provisions routes to the service network on `eth0` (pasting some NAD + pod configuration below to reproduce this use case) :

```sh
---
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: db-access-net
  namespace: default
spec:
    config: '{
        "cniVersion": "0.3.1",
        "name": "data-net",
        "netAttachDefName": "default/db-access-net",
        "topology": "localnet",
        "subnets": "192.168.200.0/24",
        "excludeSubnets": "192.168.200.1",
        "type": "ovn-k8s-cni-overlay"
    }'
---
apiVersion: v1
kind: Pod
metadata:
  name: tinywiny
  annotations:
    k8s.v1.cni.cncf.io/networks: '[
      {
        "name": "db-access-net",
        "default-route": ["192.168.200.1"]
      }
    ]'
  labels:
    role: db-adapter
spec:
  containers:
  - name: tinywiny
    image: k8s.gcr.io/e2e-test-images/agnhost:2.36 
    command: ["/bin/ash", "-c", "trap : TERM INT; sleep infinity & wait"]
    securityContext:
      runAsUser: 1000
      privileged: false
      seccompProfile:
        type: RuntimeDefault
      capabilities:
        drop: ["ALL"]
      runAsNonRoot: true
      allowPrivilegeEscalation: false

$ kubectl exec tinywiny -- ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0@if18: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP group default 
    link/ether 0a:58:0a:f4:01:05 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.244.1.5/24 brd 10.244.1.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::858:aff:fef4:105/64 scope link 
       valid_lft forever preferred_lft forever
3: net1@if19: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP group default 
    link/ether 0a:58:c0:a8:c8:02 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 192.168.200.2/24 brd 192.168.200.255 scope global net1
       valid_lft forever preferred_lft forever
    inet6 fe80::858:c0ff:fea8:c802/64 scope link 
       valid_lft forever preferred_lft forever

$ kubectl exec tinywiny -- ip r
default via 192.168.200.1 dev net1 
10.96.0.0/16 via 10.244.1.1 dev eth0 
10.244.0.0/16 via 10.244.1.1 dev eth0 
10.244.1.0/24 dev eth0 proto kernel scope link src 10.244.1.5 
100.64.0.0/16 via 10.244.1.1 dev eth0 
192.168.200.0/24 dev net1 proto kernel scope link src 192.168.200.2 
```

Reply:
With option 2 we can do that. Just a single service route like you have above going via eth1. But with option1, we are splitting the service CIDR between the default network and the user-defined primary network. So there has to be something like this (let's assume KAPI is 10.96.0.1, and DNS is 10.96.0.2):

```
$ kubectl exec tinywiny -- ip r
default via 192.168.200.1 dev net1 
10.96.0.0/16 via 10.245.1.1 dev eth1
10.96.0.1/32 via 10.244.1.1 dev eth0
10.96.0.2/32 via 10.244.1.1 dev eth0
10.244.0.0/16 via 10.244.1.1 dev eth0 
10.244.1.0/24 dev eth0 proto kernel scope link src 10.244.1.5 
10.244.5.0/24 dev eth1 proto kernel scope link src 10.245.1.5 
100.64.0.0/16 via 10.244.1.1 dev eth0
192.168.200.0/24 dev net1 proto kernel scope link src 192.168.200.2
```

You just made me realize something though while I was looking at your routes. The default cluster network join subnet 100.64.0.0/16 cannot overlap with the user-defined join subnet. Other user-defined join subnets can overlap. I'll update the doc with this.